File: yuusti/amyfunc.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <tuple>

using namespace std;

#define X first
#define Y second

using number = long long;
using P = pair<number, number>;

P sub(P p, P q) {
    return make_pair(p.X - q.X, p.Y - q.Y);
}

number dot(P p, P q) {
    return p.X * q.X + p.Y * q.Y;
}

number cross(P p, P q) {
    return p.X * q.Y - p.Y * q.X;
}

number ccw (P p, P q, P r) {
    return cross(sub(q, p), sub(r, p));
}

bool is_point_inside(vector<P> hole, P point) {
    vector<number> xs;
    for (auto p: hole) {
        xs.push_back(p.X);
    }
    number min_x, max_x;
    tie(min_x, max_x) = minmax(xs.begin(), xs.end());
    P outer = make_pair(max_x + (max_x - min_x) * 2 + 1, point.Y + 1);

    int crossings = 0;
    for (int i = 0; i < hole.size(); ++i) {
        int j = (i + 1) % hole.size();
        P d1 = sub(hole[i], point);
        P d2 = sub(hole[j], point);
        number area = cross(d1, d2);
        if (area == 0 && d1.X * d2.X && d1.Y * d2.Y <= 0) return true;
        if (ccw(point, outer, hole[i]) * ccw(point, outer, hole[j]) < 0 && ccw(hole[i], hole[j], point) * ccw(hole[i], hole[j], outer) < 0) {
            ++crossings;
        }
    }
    return crossings % 2 == 1;
}

bool is_edge_inside(vector<P> hole, vector<P> edge) {
    for (P point: edge) {
        if (!is_point_inside(hole, point)) return false;
    }
    for (int i = 0; i < hole.size(); ++i) {
        int j = (i + 1) % hole.size();
        if (ccw(edge[0], edge[1], hole[i]) * ccw(edge[0], edge[1], hole[j]) < 0 && ccw(hole[i], hole[j], edge[0]) * ccw(hole[i], hole[j], edge[1]) < 0)
            return false;
    }
    P double_mid = make_pair(edge[0].X + edge[1].X, edge[0].Y + edge[1].Y);
    vector<P> double_hole = vector<P>(hole.size());
    transform(hole.begin(), hole.end(), double_hole.begin(), [](P p){ return make_pair(p.X * 2, p.Y * 2);});

    return is_point_inside(double_hole, double_mid);
}

number d(P p, P q) {
    return (p.X - q.X) * (p.X - q.X) + (p.Y - q.Y) * (p.Y - q.Y);
}

bool is_valid_edge(P orig_p, P orig_q, P dest_p, P dest_q, number epsilon) {
    number orig_dist = d(orig_p, orig_q);
    number dest_dist = d(dest_p, dest_q);
    return abs(dest_dist - orig_dist) * 1000000 <= epsilon * orig_dist;
}

number dislike(vector<P> hole, vector<P> positions) {
    number ds = 0;
    for (P h: hole) {
        number min_p = 1e18;
        for (P p: positions) {
            min_p = min(min_p, d(h, p));
        }
        ds += min_p;
    }
}

File: ts-lib/amyfunc.ts
export interface Point {
    x: number;
    y: number;
}

export interface Edge {
    src: Point;
    dst: Point;
}

export type Vector = Point;
export type Polygon = Point[];

export function sub(p: Point, q: Point): Point {
    return {
        x: p.x - q.x,
        y: p.y - q.y
    };
}

export function dot(p: Point, q: Point): number {
    return p.x * q.x + p.y * q.y;
}

export function cross(p: Point, q: Point): number {
    return p.x * q.y - p.y * q.x;
}

export function ccw(p: Point, q: Point, r: Point): number {
    return cross(sub(q, p), sub(r, p));
}

export function isPointInside(hole: Polygon, point: Point): boolean {
    const xs = hole.map((p) => p.x);
    const maxX = xs.reduce((a, b) => Math.max(a, b));
    const minX = xs.reduce((a, b) => Math.min(a, b));
    const outer: Point = {
        x: maxX + (maxX - minX) * 2 + 1,
        y: point.y + 1
    };

    let crossings = 0;
    for (let i = 0; i < hole.length; ++i) {
        const next = (i + 1) % hole.length;
        const p1 = hole[i];
        const p2 = hole[next];
        const d1 = sub(p1, point);
        const d2 = sub(p2, point);
        const area = cross(d1, d2);
        if (area == 0 && d1.x * d2.x <= 0 && d1.y * d2.y <= 0) {
            // On this edge.
            return true
        }
        if (ccw(point, outer, p1) * ccw(point, outer, p2) < 0 && ccw(p1, p2, point) * ccw(p1, p2, outer) < 0) {
            ++crossings;
        }
    }
    return crossings % 2 == 1;
}

export function isEdgeInside(hole: Polygon, edge: Edge): boolean {
    if (!isPointInside(hole, edge.src)) {
        return false;
    }
    if (!isPointInside(hole, edge.dst)) {
        return false;
    }

    const isCrossing  = hole.some((p1, idx) => {
        const p2 = hole[(idx + 1) % hole.length];
        return ccw(edge.src, edge.dst, p1) * ccw(edge.src, edge.dst, p2) < 0 && ccw(p1, p2, edge.src) * ccw(p1, p2, edge.dst) < 0;
    });
    if (isCrossing) {
        return false;
    }

    const doubleMid: Point = {
        x: edge.src.x + edge.dst.x,
        y: edge.src.y + edge.dst.y,
    };
    const doubleHole = hole.map((p) => ({x: p.x*2, y: p.y*2}));
    if (!isPointInside(doubleHole, doubleMid)) {
        return false;
    }

    return true;
}

export function d(p: Point, q: Point): number {
    return dot(sub(p, q), sub(p, q));
}

export function isValidEdge(src: Edge, dst: Edge, eps: number): boolean {
    const srcDist = d(src.src, src.dst);
    const dstDist = d(dst.src, dst.dst);
    return Math.abs(dstDist - srcDist) * 1_000_000 <= eps * srcDist;
}

export function dislike(hole: Polygon, ps: Point[]): number {
    let sum = 0;
    hole.forEach((h) => {
        sum += ps.reduce((acc, p) => Math.min(acc, d(h, p)), Infinity);
    });
    return sum;
}
